tmpl/blog/mirage-entropy.md

```diff
@@ -0,0 +1,362 @@
+
+Before taking a look at how we try to solve this problem, let's instead consider
+what happens if we fail to do so. There is even a [Wikipedia
+page][wiki-random-attacks] on this problem, which is a nice starting point. Some
+of the highlights:
+
+The first public release of Netscape's original SSL, version 2.0, was
+[broken][ian-goldberg-netscape] several months after its release. The weakness
+was in seeding the generator with the current time, and process ID and parent
+process ID of the browser. The time stamp can be guessed to a certain precision,
+leaving only its sub-second part and the two PIDs unknown. This relatively small
+unknown space of initial values can be brute-forced.
+
+About a decade later, Debian patched their version of OpenSSL and reduced RNG
+seeding to the current PID. As a result, only 32767 random sequences were
+possible. This flaw went undetected for two years and became known as the
+[Debian fiasco][debian-fiasco]. Personal reports indicate that some of the 32767
+distinct secret keys that could be generated with OpenSSL on a Debian system
+during that time are still in circulation.
+
+Computing the largest common divisor of a pair of numbers is much faster than
+discovering all the prime divisors of a particular number. RSA public keys
+contain a number, and secrets keys contain its factors. An RSA key is usually
+generated by randomly picking the factors. If a pool of keys was generated with
+a heavily biased random number generator, such that factors are likely to
+repeat, it is possible to search for common factors in all pairs and crack the
+affected keys, a technique which produces
+[spectacular][factoring-rsa-smartcards] [results][mining-your-ps-and-qs].
+
+Recently, a bitcoin application for Android was
+[discovered][reddit-bitcoin-post] to be downloading its random seed from a
+[website](http://www.random.org). It wasn't even necessary to intercept this
+unencrypted traffic, because the website started serving a redirect page and the
+Android application was left seeding its RNG with the text of the redirection
+message. It therefore started
+generating the same private ECDSA key and the associated bitcoin address for
+every affected user, an issue which reportedly [cost][thereg-bitcoin] some users
+their bitcoins.
+
+Playstation 3 game signatures can be forged. Sony [reused][schneier-sony] a
+single `k`-parameter, which is supposed to be "unique, unpredictable and
+secret", for every ECDSA signature they made. This lead to complete compromise
+of the signing keys. Admittedly, this not really a random number generator
+problem in itself, but it is a limit situation of an RNG malfunction.
+
+These only are some of the most spectacular failures related to random numbers.
+For example, it is widely known in security circles that RNGs of embedded
+devices tend to be predictable, leading for example to widespread use of weak
+key on routers and similar equipment. So if you were designing a unikernel operating
```

Comment:
**So when implementing a unikernel operating system, you don't want to end up on that Wikipedia page either.**